// @sitecore-jss/sitecore-jss/dist/cjs/utils/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetExperienceEditorChromes = exports.isExperienceEditorActive = exports.resetEditorChromes = exports.isEditorActive = exports.HorizonEditor = exports.ExperienceEditor = exports.resolveUrl = exports.isServer = void 0;
const is_server_1 = __importDefault(require("./is-server"));
exports.isServer = is_server_1.default;
const resolve_url_1 = __importDefault(require("./resolve-url"));
exports.resolveUrl = resolve_url_1.default;
var editing_1 = require("./editing");
Object.defineProperty(exports, "ExperienceEditor", { enumerable: true, get: function () { return editing_1.ExperienceEditor; } });
Object.defineProperty(exports, "HorizonEditor", { enumerable: true, get: function () { return editing_1.HorizonEditor; } });
Object.defineProperty(exports, "isEditorActive", { enumerable: true, get: function () { return editing_1.isEditorActive; } });
Object.defineProperty(exports, "resetEditorChromes", { enumerable: true, get: function () { return editing_1.resetEditorChromes; } });
Object.defineProperty(exports, "isExperienceEditorActive", { enumerable: true, get: function () { return editing_1.isExperienceEditorActive; } });
Object.defineProperty(exports, "resetExperienceEditorChromes", { enumerable: true, get: function () { return editing_1.resetExperienceEditorChromes; } });
